### PR TITLE
Niro/feat/auth

### PR DESCRIPTION
**Pull request by niroshanwitharana**

### CODE REVIEW

File: db/models/user.js [hunk 1]
```diff
@@ -1,12 +1,12 @@
 const mongoose = require("mongoose");
+const Schema = require('mongoose').Schema
```
Comment (JoeDodgson) [inline db/models/user.js:2]: This can be changed to `const Schema = mongoose.Schema;`

File: db/models/user.js [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+const bcrypt = require('bcrypt');
 
-const Schema = mongoose.Schema;
-
-const UserSchema = new Schema({
```
Comment (JoeDodgson) [inline db/models/user.js:5]: Why did you remove this?

File: db/models/user.js [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
+const UserSchema = new mongoose.Schema({
     email: {
         type: String,
         unique: true,
-        match: [/.+@.+\..+/, "Please enter a validate email"]
+         match: [/.+@.+\..+/, "Please enter a validate email"]
```
Comment (JoeDodgson) [inline db/models/user.js:9]: Typo- should be "Please enter a *valid* email"

File: db/models/user.js [hunk 4]
```diff
@@ -25,18 +25,56 @@ const UserSchema = new Schema({
...
+            return cb(err);
+            // if there are no errors with compareing
+            else{
+                // if passwords match?
+                if(!isMatch)
+                // is not match call done function isMatch will be false
+                return cb(null, isMatch);
+                // if there is a match call done and return user object
+                // "this" mean user 
+                return cb(null, this);
```
Comment (JoeDodgson) [inline db/models/user.js:72]: Will this return statement ever get executed? because it is after the return statement on line 69

File: middleware/passport.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        return done(null, user);
+        else
+        // there is no user has that primary key
+        return done(null, false)
+    })
+}))
+
+//  authenticated local strategy using email and passwod
+passport.use(new localStrategy((username, password, done) =>{
+    console.log(username);
```
Comment (JoeDodgson) [inline middleware/passport.js:41]: Should we be console logging username here? If yes, maybe console log something a bit more readable like
```console.log(`${username} has been authenticated successfully`);```

File: routes/User.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
+const express = require("express");
+const userRouter = express.Router();
+const passport = require("passport");
+const passportConfig = require("../middleware/passport");
+// require('dotenv').config();
```
Comment (JoeDodgson) [inline routes/User.js:5]: I assume we do need this because we are accessing config variables from process.env (e.g. on line 16)

File: routes/User.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  }
+);
+
+//Creating the logout route
+userRouter.get(
+  "/logout",
+  passport.authenticate("jwt", { session: false }),
+  (req, res) => {
+    res.clearCookie("access_token");
+    res.json({ user: { emai: "" }, success: true });
```
Comment (JoeDodgson) [inline routes/User.js:97]: Is this a typo? Should the property be *email*?

**Review by JoeDodgson — CHANGES_REQUESTED** [review]
Really good work Niro. Just some minor comments to address

File: client/src/Services/AuthService.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    logout : () => {
+        return fetch('/user/logout')
+            .then(res => res.json())
+            .then(data => data);
+    },
+    // this isAuthenticated function use to persist authentican
+    // once user login state in the react app will know user has been authenticated, but when user close the app the state will gone.
+    // this function will sinc front-end and back-end and keep user authenticated even react app is closed.
+    // so, when user visit the website next time user will still stay login
+    //we use context-API to call this function, it is a global state for our react app
```
[the remark this answers is not in the document]
Comment (niroshanwitharana) [reply client/src/Services/AuthService.js:34]: yep i'm confident about it

File: .env [hunk 9]
```diff
@@ -0,0 +1,4 @@
+DB_PASSWORD=yLGEHfsCY4nx3VJm
+MONGODB_URI=mongodb+srv://niroshan:yLGEHfsCY4nx3VJm@niro-passport.3r929.mongodb.net/engineerdb?retryWrites=true&w=majority
+SECRET_OR_KEY=riskManager
+JWT_ISSUE=riskManager-Admin
```
[the remark this answers is not in the document]
Comment (niroshanwitharana) [reply .env]: changes done as required

File: db/models/user.js (hunk 1 above)
Comment (niroshanwitharana) [reply db/models/user.js:2]: changes done as required

File: db/models/user.js (hunk 2 above)
Comment (niroshanwitharana) [reply db/models/user.js:5]: there should be required(bcrypt), resolve it

File: db/models/user.js (hunk 3 above)
Comment (niroshanwitharana) [reply db/models/user.js:9]: i did changed it. DB-Schema stuff coming from Ian ;)

File: db/models/user.js (hunk 4 above)
Comment (niroshanwitharana) [reply db/models/user.js:72]: both return statement will get executed according to password is match or not, 69 will be under the if statement(if password isMatch false) and 72 will execute under the else statement(when password isMatch is true)

File: middleware/passport.js (hunk 5 above)
Comment (niroshanwitharana) [reply middleware/passport.js:41]: been using for debugging, will no need further more

File: routes/User.js (hunk 7 above)
Comment (niroshanwitharana) [reply routes/User.js:97]: here we responding user info as an object, will need to add some more info after checked with the front-end

File: .env (hunk 9 above)
[the remark this answers is not in the document]
Comment (JoeDodgson) [reply .env]: I'm happy that this has been addressed

File: db/models/user.js (hunk 1 above)
Comment (JoeDodgson) [reply db/models/user.js]: I'm happy that this has been addressed

File: db/models/user.js (hunk 3 above)
Comment (JoeDodgson) [reply db/models/user.js]: I'm happy that this has been addressed

File: db/models/user.js (hunk 2 above)
Comment (JoeDodgson) [reply db/models/user.js]: I'm happy that this has been addressed

File: db/models/user.js (hunk 4 above)
Comment (JoeDodgson) [reply db/models/user.js:72]: Ok that makes sense

File: middleware/passport.js (hunk 5 above)
Comment (JoeDodgson) [reply middleware/passport.js]: I'm happy that this has been addressed

File: routes/User.js (hunk 6 above)
Comment (JoeDodgson) [reply routes/User.js]: I'm happy that this has been addressed

File: routes/User.js (hunk 7 above)
Comment (JoeDodgson) [reply routes/User.js]: I'm happy that this has been addressed

**Review by JoeDodgson — APPROVED** [review]
I'm happy that all comments have been addressed
